### PR TITLE
Signal tokens moved to exchange

### PR DESCRIPTION
**Pull request by IvanIvanoff**
#### Summary

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: lib/sanbase/notifications/behaviour.ex [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  @typedoc ~s"""
+  String describing the place where the notification is published.
+  Currently we have implementations for "discord" and "discourse"
+  """
+  @type publish_place :: String.t()
+
+  @type payload :: any
+
+  @callback run :: :ok | {:error, String.t()}
+  @callback publish(publish_place, payload) :: :ok | {:error, String.t()}
```
[the remark this answers is not in the document]
Comment (tspenov) [reply lib/sanbase/notifications/behaviour.ex]: @IvanIvanoff  I think next we might need `run` taking the `user` and as param and based on user signal preferences - calculates all signals.
```
    user_signals: [
      schedule: "@daily",
      task: {Sanbase.Notifications.Users, :run, []}
    ]
```
In Sanbase.Notifications.Users we take all users, loop and check their signal prefs, and call DaaSignal.run(user, user_signal_prefs), ExchangeFlowSignal.run(user, user_signal_prefs) ,..., calculate payload from all and send one email. 
But we can start with discord signals and add these on top ?

[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase/notifications/behaviour.ex]: About run accepting users, I have a kind of different idea for it, though it's too long to describe here.
Shortly, I think we should implement a central unit that calculates some metrics like percent of tokens deposited into an exchange. Only after that this recalculated result will be evaluated against different user preferences so we avoid each user querying the databases.

[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase/notifications/behaviour.ex]: Let's move the discussion to a specific card.

File: config/notifications_config.exs [hunk 2]
```diff
@@ -20,3 +20,17 @@ config :sanbase, Sanbase.Notifications.PriceVolumeDiff,
...
+  daa_signal_discord_publish_user: {:system, "DAA_SIGNAL_DISCORD_PUBLISH_USER", "DAA GOING UP"},
+  daa_signal_threshold: {:system, "DAA_SIGNAL_THRESHOLD", "100"},
+  daa_signal_timeframe_from: {:system, "DAA_SIGNAL_TIMEFRAME_FROM", "30"},
+  daa_signal_timeframe_to: {:system, "DAA_SIGNAL_TIMEFRAME_TO", "2"},
+  daa_signal_change: {:system, "DAA_SIGNAL_CHANGE", "3"}
+
+config :sanbase, Sanbase.Notifications.Discord.ExchangeInflow,
+  webhook_url: {:system, "EXCHANGE_INFLOW_DISCORD_WEBHOOK_URL"},
+  publish_user: {:system, "DAA_SIGNAL_DISCORD_PUBLISH_USER", "DAA GOING UP"},
+  signal_trigger_percent: {:system, "EXCHANGE_INFLOW_TRIGER_PERCEN", "1"},
```
Comment (tspenov) [inline config/notifications_config.exs]: `PERCENT`

File: config/notifications_config.exs [hunk 3]
```diff
@@ -20,3 +20,17 @@ config :sanbase, Sanbase.Notifications.PriceVolumeDiff,
...
+  webhook_url: {:system, "DAA_SIGNAL_DISCORD_WEBHOOK_URL"},
+  daa_signal_discord_publish_user: {:system, "DAA_SIGNAL_DISCORD_PUBLISH_USER", "DAA GOING UP"},
+  daa_signal_threshold: {:system, "DAA_SIGNAL_THRESHOLD", "100"},
+  daa_signal_timeframe_from: {:system, "DAA_SIGNAL_TIMEFRAME_FROM", "30"},
+  daa_signal_timeframe_to: {:system, "DAA_SIGNAL_TIMEFRAME_TO", "2"},
+  daa_signal_change: {:system, "DAA_SIGNAL_CHANGE", "3"}
+
+config :sanbase, Sanbase.Notifications.Discord.ExchangeInflow,
+  webhook_url: {:system, "EXCHANGE_INFLOW_DISCORD_WEBHOOK_URL"},
+  publish_user: {:system, "DAA_SIGNAL_DISCORD_PUBLISH_USER", "DAA GOING UP"},
```
Comment (tspenov) [inline config/notifications_config.exs]: the user has to be changed

File: lib/sanbase/notifications/discord/exchange_inflow.ex [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      Timex.shift(Timex.now(), days: -interval_days()),
+      Timex.now()
+    )
+    |> case do
+      {:ok, list} ->
+        publish("discord", build_payload(projects, list))
+
+      {:error, error} ->
+        Logger.error("Error getting exchange inflowfrom TimescaleDB. Reason: #{inspect(error)}")
+        з{:error, "Error getting exchange inflowfrom TimescaleDB."}
```
Comment (tspenov) [inline lib/sanbase/notifications/discord/exchange_inflow.ex]: * some leftover char before `{`
* inflowfrom -> `inflow from`

File: lib/sanbase/notifications/discord/exchange_inflow.ex [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    projects
+    |> Enum.map(fn %Project{main_contract_address: contract} -> contract end)
+    |> Sanbase.Blockchain.ExchangeFundsFlow.transactions_in(
+      Timex.shift(Timex.now(), days: -interval_days()),
+      Timex.now()
+    )
+    |> case do
+      {:ok, list} ->
+        publish("discord", build_payload(projects, list))
```
Comment (tspenov) [inline lib/sanbase/notifications/discord/exchange_inflow.ex]: * I think the payload should be in json format in the end for discord
* if there is nothing to signal (empty content) - nothing should be published to discord

File: lib/sanbase/notifications/discord/exchange_inflow.ex [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+      Enum.map(list, fn %{contract: contract, inflow: inflow} -> {contract, inflow} end)
+      |> Map.new()
+
+    projects
+    |> Enum.map(fn %Project{} = project ->
+      inflow = Map.get(contract_inflow_map, project.main_contract_address)
+
+      if percent_of_total_supply(project, inflow) > signal_trigger_percent() do
+        """
+        Project #{project.coinmarketcap_id} (contract address #{project.main_contract_address} has more
```
[the remark this answers is not in the document]
Comment (IvanIvanoff) [reply lib/sanbase/notifications/discord/exchange_inflow.ex]: Currently we do not show the exchange flow on sanbase, though it is in the API. Nevertheless, I will add the link

File: lib/sanbase/prices/store.ex [hunk 7]
```diff
@@ -212,4 +238,42 @@ defmodule Sanbase.Prices.Store do
...
+    combined_volume = values |> Enum.reduce(0, fn [[_, vol, _]], acc -> acc + vol end)
+    combined_mcap = values |> Enum.reduce(0, fn [[_, _, mcap]], acc -> acc + mcap end)
+    marketcap_values = values |> Enum.map(fn [[_, _, mcap]] -> mcap end)
+
+    marketcap_percent =
+      marketcap_values |> Enum.map(fn mcap -> Float.round(mcap / combined_mcap, 5) end)
+
+    slug_marketcap_percent_map = Enum.zip(slugs, marketcap_percent) |> Enum.into(%{})
+
+    {:ok, combined_volume, combined_mcap, slug_marketcap_percent_map}
```
Comment (tspenov) [inline lib/sanbase/prices/store.ex]: hmm, this seems old code: plese check the version in master: Sanbase.Notifications.Discord.DaaSignal.run()

**Pull request merged**